libres/lib/res_util/matrix.cpp

```diff
@@ -1,556 +1,56 @@
-                         const char *fmt) {
-    matrix_pretty_fprint(matrix, name, fmt, stdout);
-}
-
-void matrix_fprintf(const matrix_type *matrix, const char *fmt, FILE *stream) {
-    int i, j;
-    for (i = 0; i < matrix->rows; i++) {
-        for (j = 0; j < matrix->columns; j++)
-            fprintf(stream, fmt, matrix_iget(matrix, i, j));
-        fprintf(stream, "\n");
-    }
-}
-
-void matrix_dump_csv(const matrix_type *matrix, const char *filename) {
-    FILE *stream = util_fopen(filename, "w");
-    for (int i = 0; i < matrix->rows; i++) {
-        for (int j = 0; j < matrix->columns - 1; j++)
-            fprintf(stream, "%g, ", matrix_iget(matrix, i, j));
-        fprintf(stream, "%g\n", matrix_iget(matrix, i, matrix->columns - 1));
-    }
-    fclose(stream);
-}
-
-void matrix_fwrite(const matrix_type *matrix, FILE *stream) {
-    util_fwrite_int(matrix->rows, stream);
-    util_fwrite_int(matrix->columns, stream);
-
-    if (matrix->column_stride == matrix->rows)
-        util_fwrite(matrix->data, sizeof *matrix->data,
-                    matrix->columns * matrix->rows, stream, __func__);
-    else {
-        int column;
-        for (column = 0; column < matrix->columns; column++) {
-            if (matrix->row_stride == 1) {
-                const double *column_data =
-                    &matrix->data[column * matrix->column_stride];
-                util_fwrite(column_data, sizeof *column_data, matrix->rows,
-                            stream, __func__);
-            } else {
-                int row;
-                for (row = 0; row < matrix->rows; row++)
-                    util_fwrite_double(
-                        matrix->data[GET_INDEX(matrix, row, column)], stream);
-            }
-        }
-    }
-}
-
-void matrix_fread(matrix_type *matrix, FILE *stream) {
-    int rows = util_fread_int(stream);
-    int columns = util_fread_int(stream);
-
-    matrix_resize(matrix, rows, columns, false);
-    if (matrix->column_stride == matrix->rows)
-        util_fread(matrix->data, sizeof *matrix->data,
-                   matrix->columns * matrix->rows, stream, __func__);
-    else {
-        int column;
-        for (column = 0; column < matrix->columns; column++) {
-            if (matrix->row_stride == 1) {
-                double *column_data =
-                    &matrix->data[column * matrix->column_stride];
-                util_fread(column_data, sizeof *column_data, matrix->rows,
-                           stream, __func__);
-            } else {
-                int row;
-                for (row = 0; row < matrix->rows; row++)
-                    matrix->data[GET_INDEX(matrix, row, column)] =
-                        util_fread_double(stream);
-            }
-        }
-    }
-}
-
-matrix_type *matrix_fread_alloc(FILE *stream) {
-    matrix_type *matrix = matrix_alloc(1, 1);
-    matrix_fread(matrix, stream);
-    return matrix;
+    if (matrix)
```

Comment:
This is because of the `nullptr`?

Reply:
This is because we had `matrix_safe_free` I merged into this function. However, I checked the standard and both `free()` and `delete` are guaranteed do nothing when `nullptr` is passed, so this `if` is unnecessary.